File: AULA_07/Haskell/Matriz.hs
import Data.List.Split
import Data.List
import System.IO

somaMatriz m1 m2 = zipWith (\a b -> zipWith (+) a b) m1 m2

transposta :: [[a]] -> [[a]]
transposta ([]:_) = []
transposta x = (map head x) : transposta (map tail x)

prodExt v1 v2 = map (\x -> map (\y -> x*y) v2) v1

main = do
        --f <- openFile "matriz1.csv" ReadMode
        --linhas <- hGetContents f
        linhas1 <- readFile "matriz1.csv"
        linhas2 <- readFile "matriz2.csv"
        linhas3 <- readFile "vetor1.csv"
        linhas4 <- readFile "vetor2.csv"
        let matrizStr x = map (splitOn ";") (lines x)
            matriz1 = map (map read) (matrizStr linhas1) :: [[Int]]
            matriz2 = map (map read) (matrizStr linhas2) :: [[Int]]
            vetor1  = map read (splitOn ";" linhas3) :: [Int]
            vetor2  = map read (splitOn ";" linhas4) :: [Int]
        putStrLn $ show matriz1
        putStrLn $ show matriz2
        putStrLn $ show vetor1 
        putStrLn $ show vetor2
        let matriz3 = somaMatriz matriz1 matriz2
            matrizT = transpose matriz1
            matriz4 = prodExt vetor1 vetor2
        putStrLn $ show matriz3
        putStrLn $ show matrizT
        putStrLn $ show matriz4

File: AULA_07/C/Matriz.c
#include <stdio.h>
#include <stdlib.h>

int main(){

  double **A, **B, **C, **T, **D;
  double *x, *y;
  FILE * f;
  int n1,m1, n2,m2, n3, n4;
  int i,j;

  f = fopen("matriz1.csv","r");
  fscanf(f,"%d;%d",&n1,&m1);

  A = (double **) malloc(n1*sizeof(double *));
  for( i=0;i<n1;i++ ) A[i]=(double *) malloc(m1*sizeof(double));

  for( i=0;i<n1;i++ )
    for( j=0;j<m1;j++ )
      fscanf(f,"%lf;",&A[i][j]);
  fclose(f);

  f = fopen("matriz2.csv","r");
  fscanf(f,"%d;%d",&n2,&m2);

  B = (double **) malloc(n1*sizeof(double *));
  for( i=0;i<n2;i++ ) B[i] = (double *) malloc(m2*sizeof(double));

  for( i=0;i<n2;i++ )
    for( j=0;j<m2;j++ )
      fscanf(f,"%lf;",&B[i][j]);
  fclose(f);

  f = fopen("vetor1.csv","r");
  fscanf(f,"%d",&n3);
  x = (double *) malloc(n3*sizeof(double));
  for( i=0;i<n3;i++ ) fscanf(f,"%lf;",&x[i]);
  fclose(f);

  f = fopen("vetor2.csv","r");
  fscanf(f,"%d",&n4);
  y = (double *) malloc(n4*sizeof(double));
  for( i=0;i<n4;i++ ) fscanf(f,"%lf;",&y[i]);
  fclose(f);

  /* C = A + B */
  if( n1==n2 && m1==m2 ){
    C = (double **) malloc(n1*sizeof(double *));
    for( i=0;i<n1;i++ ) C[i] = (double *) malloc(m1*sizeof(double));

    for( i=0;i<n1;i++ )
      for( j=0;j<m1;j++ )
        C[i][j] = A[i][j] + B[i][j];

    f = fopen("somaMatriz.csv","w+");
    fprintf(f, "%d;%d\n", n1,m1);
    for( i=0;i<n1;i++ ){
      for( j=0;j<m1-1;j++ ){
        fprintf(f, "%lf;", C[i][j]);
      }
      fprintf(f, "%lf\n", C[i][j]);
    }
    fclose(f);
  }

  /* T = A' */
  T = (double **) malloc(m1*sizeof(double *));
  for( i=0;i<m1;i++ ) T[i] = (double *) malloc(n1*sizeof(double));

  for( i=0;i<n1;i++ )
    for( j=0;j<m1;j++ )
      T[j][i] = A[i][j];

  f = fopen("transposta.csv","w+");
  fprintf(f, "%d;%d\n", m1,n1);
  for( i=0;i<m1;i++ ){
    for( j=0;j<n1-1;j++ ){
      fprintf(f, "%lf;", T[i][j]);
    }
    fprintf(f, "%lf\n", T[i][j]);
  }
  fclose(f);

  /* x o y = D */
  D = (double **) malloc(n3*sizeof(double *));
  for( i=0;i<n3;i++ ) D[i] = (double *) malloc(n4*sizeof(double));

  for( i=0;i<n3;i++ )
    for( j=0;j<n4;j++ )
      D[i][j] = x[i]*y[j];

  f = fopen("produtoexterno.csv","w+");
  fprintf(f, "%d;%d\n", n3,n4);
  for( i=0;i<n3;i++ ){
    for( j=0;j<n4-1;j++ ){
      fprintf(f, "%lf;", D[i][j]);
    }
    fprintf(f, "%lf\n", D[i][j]);
  }
  fclose(f);

  return 0;
}
